### DISCUSSION TITLE
How to apply migrations to tests?

### DISCUSSION THREAD

**Discussion by DigNZ** [pocketbase/pocketbase#396] [Q&A] [answered]
My code make Dao() calls for things like getting collections. I've used the command line tool to make collections migration and checked that it works successfully when running a new install of the application - creating all the collections I except correctly. 

However tests all fail with errors as they work on a test database and the migrations are not automatically applied to that database. I added a step in my test start up in order to apply migrations:

```
	initTestApp := func(t *testing.T, app *tests.TestApp) {
		runner, err := migrate.NewRunner(
			app.DB(),
			migrations.AppMigrations,
		)
		if err != nil {
			t.Error(err)
		}
		applied, err := runner.Up()
		if err != nil {
			t.Error(err)
		}
		t.Logf("Applied: %v", applied)
		
	}
```
However

I get this error:
`Failed to apply migration 1662546327_collections_snapshot.go: System collection "profiles" cannot be deleted.` 

and an empty array of applied migrations.

How can I apply migrations to the test data so that my application code can see expected collections? Or should it be handled in a different way?

**Comment by ganigeorgiev** [comment, marked as answer]
I will have to add support for loading an external test database in order to run local migrations in the tests via the `tests.TestApp` helper.

The reason for the error is that the exported collections in your migration file have ids that differ from those created during the test and any missing collection from the `collections_snapshot` is treated as deleted.

The test helpers and the documentation will be revisited more thoroughly with the ongoing refactoring in https://github.com/pocketbase/pocketbase/issues/376.

Since there is no easy solution for this without writing your own helper, I'll try to implement some of the required changes in the test helpers for the next maintenance release I'm planning this week (probably it will be the last release before the breaking changes from #376).

**Reply by ganigeorgiev** [reply]
@DigNZ Support to allow external test data dir for the `tests.TestApp` instance was added in master.

I've also update the `tests.ApiScenario` to include a `TestAppFactory` hook which can be used to both provide your test app instance and init your custom api routes.

You can manually run the migration as in your example OR create a full `test_pb_data` directory (including test storage files, not just db schema). The main benefit of a full test data directory is that you'll be able to setup your test data with the Admin UI, by just starting the application pointing to the test data dir, eg.:
```go
./go run main.go serve --dir="./test_pb_data/"
# open 127.0.0.1:8090 and setup your test data
```
After you are done with the setup of your test data, you can pass the directory path to the TestAppFactory like this:

```go
package main

import (
	"net/http"
	"testing"

	"github.com/pocketbase/pocketbase/tests"
)

func TestHelloEndpoint(t *testing.T) {
	// setup the test app instance
	setupTestApp := func() (*tests.TestApp, error) {
		app, err := tests.NewTestApp("./test_pb_data")
		if err != nil {
			return nil, err
		}

		bindAppHooks(app)

		return app, nil
	}

	scenarios := []tests.ApiScenario{
		{
			Name:            "try with different http method, eg. POST",
			Method:          http.MethodPost,
			Url:             "/my/hello",
			ExpectedStatus:  405,
			ExpectedContent: []string{"\"data\":{}"},
			TestAppFactory:  setupTestApp,
		},
		// ... other tests ...
	}

	for _, scenario := range scenarios {
		scenario.Test(t)
	}
}
```

The test directory is safe to be modified by the tests because it will be fully cloned to a temporary location when initializing the test app and will be removed on `testApp.Cleanup()` (it is called automatically by `tests.ApiScenario`).

Sometime tomorrow I'll try to prepare a release with the changes (v0.7.0).

_Please note that other changes to the tests helpers and proper tests documentation will be done together/after #376._

**Reply by DigNZ** [reply]
That is excellent, thanks for the quick turn around!
